rossum/lib/api_client.py

```diff
@@ -219,30 +219,62 @@ def get_paginated(
         query: Optional[Dict[str, Any]] = None,
         *,
         key: str = "results",
+        sideloads: Optional[Iterable[APIObject]] = None,
     ) -> Tuple[List[Dict[str, Any]], int]:
+        if query is None:
+            query = {}
+
+        if sideloads is not None:
+            if "sideload" in query:
+                raise RossumException("sideloading cannot be specified both in query and sideloads")
+            query = query.copy()
+            for sideload in sideloads:
+                sideload.setup_sideload_query(query)
+
         response = self.get(path, query)
         response_dict = response.json()
 
-        res = response_dict[key]
+        res = response_dict
         next_page = response_dict["pagination"]["next"]
 
         while next_page:
             response = self.get_url(next_page)
             response_dict = response.json()
 
-            res.extend(response_dict[key])
```

Comment:
I think that the reason why we choose the values from the key here rather than at the end is performance. When listing annotations, you may be listing hundreds thousands of objects, which takes quite a bit of time and other resources. Such objects may have really a lot of subvalues that you are trying to sideloads and storing all the objects first and then choosing just the ones under a cetain key can get much more demanding than storing the chosen values first (which imho is the most optimal approach).

Reply:
###  Sideloading Time & Memory Demands

The following describes time and memory requirements measured with `get_paginated(...)` with the same set of parameters as `get_annotations(...)` calls, both with varying sideload parameters (queues contained different number of annotations, but same schema).
Note that when content was used in sideloads, `get_paginated(..., key="content")` was used to make sensible and reasonable comparison.

Annotations: 35, Datapoints types 46, Sideload: documents, modifiers, automation_blockers, content
⌛: 2.72s 💾: 59.3 MiB
⌛: 2.78s 💾: 59.2 MiB

Annotations: 1036, Datapoints types: 46, Sideload: documents, modifiers, automation_blockers, content
⌛: 2min 15s 💾: 157.7 MiB
⌛: 2min 29s 💾: 162.7 MiB

Annotations: 25236, Datapoints types: 46, Sideload: documents, modifiers, automation_blockers, content
⌛:    57min  7s 💾: 2052.4 MiB
⌛: 1h  1min 17s 💾: 2203.1 MiB

Some more measurements for the "worst-case":

Annotations: 25236, Sideload: documents
⌛: 23min 58s 💾: 138.7 MiB
⌛: 24min 13s 💾: 185.5 MiB

Annotations: 25236, Sideload: documents, modifiers, automatin_blockers 
⌛: 22min 57s 💾: 138.6 MiB
⌛: 23min 37s 💾: 204.6 MiB

Annotations: 25236, Datapoints types: 46, Sideload: content
⌛: 1h 1min 22s 💾: 2051.5 MiB
⌛: 1h 9min  7s 💾: 2143.2 MiB

I think it looks good and despite the memory requirements go up they are in reasonable bounds 🤞.

Reply:
I agree, it looks acceptable :) Thank you very much for the measurements.